nemo/collections/tts/modules/submodules.py

```diff
@@ -410,3 +419,284 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+            "style_emb": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs):
+        bs = inputs.size(0)
+        q = self.q_linear(inputs.unsqueeze(1))
+        k = self.k_linear(self.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1))
+        v = self.v_linear(self.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1))
+
+        q = q.view(bs, q.shape[1], self.n_head, self.token_size)
+        k = k.view(bs, k.shape[1], self.n_head, self.token_size)
+        v = v.view(bs, v.shape[1], self.n_head, self.token_size)
+
+        q = q.permute(2, 0, 1, 3).contiguous().view(-1, q.shape[1], q.shape[3])
+        k = k.permute(2, 0, 3, 1).contiguous().view(-1, k.shape[3], k.shape[1])
+        v = v.permute(2, 0, 1, 3).contiguous().view(-1, v.shape[1], v.shape[3])
+
+        scores = torch.bmm(q, k) / self.temperature
+        scores = self.softmax(scores)
+
+        style_emb = torch.bmm(scores, v).squeeze(1)
+        style_emb = style_emb.contiguous().view(self.n_head, bs, self.token_size)
+        style_emb = style_emb.permute(1, 0, 2).contiguous().view(bs, -1)
+
+        return style_emb
+
+
+class SVEmbeddingLinearProjectionNetwork(torch.nn.Module):
+    def __init__(self, speaker_embedding_dim, symbols_embedding_dim):
+        super(SVEmbeddingLinearProjectionNetwork, self).__init__()
+        self.speaker_proj = torch.nn.Linear(speaker_embedding_dim, symbols_embedding_dim)
+        self.act_fn = torch.nn.ReLU()
+
+    def forward(self, sv_embedding):
+        x = self.speaker_proj(sv_embedding)
+        return x.unsqueeze(1)
+
+
+class SpeakerEncoder(torch.nn.Module):
+    """
+    class SpeakerEncoder represents speakers representation. This module can combine GST (global style token)
+    based speaker embeddings, speaker embeddings from speaker verificaiton models and lookup table speaker
+    embeddings.
+    """
+
+    def __init__(self, gst_module=None, sv_projection_module=None, lookup_emb_projection_module=None):
+        """
+        Constructor.
+        Parameters:
+            - gst_module: Neural module to get GST based speaker embedding
+            - sv_projection_module: Neural module to get Speaker Verification based speaker embedding
+            - lookup_emb_projection_module: Neural module to project lookup speaker embedding
+        """
+        super(SpeakerEncoder, self).__init__()
+        self.gst_module = gst_module
+        self.sv_projection_module = sv_projection_module
+        self.lookup_emb_projection_module = lookup_emb_projection_module
+
+    def forward(self, spk_emb, reference_spec=None, reference_spec_lens=None, speaker_embedding=None):
+        x = spk_emb
+        # Project lookup speaker embedding
+        if self.lookup_emb_projection_module is not None:
+            x += self.lookup_emb_projection_module(spk_emb)
+
+        # Get GST based speaker embedding
+        if self.gst_module is not None and reference_spec is not None and reference_spec_lens is not None:
```

Comment:
Should we throw an error if gst_module is not None but reference spec is None?

Reply:
Add the check logic in `models/fastpitch.py` https://github.com/NVIDIA/NeMo/blob/tts_adapters_cln/nemo/collections/tts/models/fastpitch.py#L151-L176